fishers-exact-test/d/dfet.d

```diff
@@ -47,51 +55,40 @@ double logHypergeometricProbability(const long[] data, const ref double[] fs) {
 		fs[data[1]] -
 		fs[data[2]] -
 		fs[data[3]] -
-		fs[data[0] + data[1] + data[2] + data[3]]
-	);
+		fs[data[0] + data[1] + data[2] + data[3]];
 }
 
+// really need to do more than this pragma to make sure the optimizer doesn't cheat
+// and inline fisherExact in to main and then start using up-front knowledge of the
+// contingency table or see that they whole "loop n times" edifice doesn't change the
+// result, or something more subtle (pulling some calculations outside of the loop),
+// all of which will give misleading benchmark results.
+pragma(inline, false)
 double fisherExact(const long[] data) {
 	// sum all table values
 	const grandTotal = data.sum;
-	
+
 	// save factorial values for repeated use in the loop below
 	const factorials = logFactorial(grandTotal);
 
 	// calculate our rejection threshold
 	const pvalThreshold = logHypergeometricProbability(data, factorials);
 
 	double pvalFraction = 0;
-	for(long i = 0; i <= grandTotal; i++) {
-		if((data[0] + data[1] - i >= 0) && (data[0] + data[2] - i >= 0) && (data[3] - data[0] + i >=0)) {
-			double lhgp = logHypergeometricProbability([
-				i,
-				data[0] + data[1] - i,
-				data[0] + data[2] - i,
-				data[3] - data[0] + i
-			], factorials);
-
-			if(lhgp <= pvalThreshold) {
-				pvalFraction += exp(lhgp - pvalThreshold);
-			}
+	foreach(i; data[3] - data[0] .. data[0] + min(data[1], data[2]) + 1) {
```

Comment:
I spent a grand total of <2 minutes thinking about this line, so I suggest you check it carefully if you want to use it!